taichi/backends/cuda/codegen_cuda.cpp

```diff
@@ -252,7 +252,16 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
       auto dst_type =
           stmt->dest->ret_type->as<PointerType>()->get_pointee_type();
       if (stmt->op_type == AtomicOpType::add) {
-        if (dst_type->is<PrimitiveType>() && is_integral(stmt->val->ret_type)) {
+        if (stmt->is_reduction &&
```

Comment:
nit: that's a lot of if-elses... Could you create a separate PR to clean this up? One way to do it would be to separate this into a few cases and use lookup tables. E.g.

```cpp
if (stmt->is_reduction) {
  old_value = handle_atomic_reduction(stmt, ...);
} else if (auto cit = dst_type->cast<CustomIntType>()) {
  old_value = atomic_add_custom_int(stmt, ...);
} else if (...) {
  // custom float
} else {
  // handle the rest of the cases...
}
```

whereas `handle_atomic_reduction()` can be:

```cpp
llvm::Value* handle_atomic_reduction(AtomicOpStmt *stmt, ...) {
  // precondition checks: |stmt| is of primitive type, op type is add, etc.

  // m := a table that maps from {PrimitiveType, AtomicOpType} --> name of the function
  return create_call(m[primite_type][atomic_op], ...);
}
```

Reply:
Sure. Will create another PR for this shortly.